### DISCUSSION TITLE
Support for ipfs:// and ipns:// protocols

### DISCUSSION THREAD

**Discussion by fabricedesre** [servo/servo#29332] [Ideas]
ipfs:// and ipns:// provide access to content based addressed resources. They are agnostic to the payload so can be used to load web content. Because of content based addressing, content integrity is guaranteed independently of the transport layer used as long as you verify block hashes.

Broadly speaking there are 2 classes of implementations:
- redirecting ipfs:// urls to a http(s) gateway. This is what browser extensions do, but is not ideal since you delegate the integrity check to the gateway, hence the trust.
- using a local IPFS node so get local integrity check. Brave can do that by running a Kubo node, and Capyloon does that by running a Iroh node (https://github.com/n0-computer/iroh).

While running a full node works, it's also quite heavyweight. A better solution would be to extract the block verification and traversal from eg. iroh's code base and use it for a lightweight implementation. This component could be generic enough to be reused in multiple browsers and other runtimes.

Is there any appetite to get such support in Servo?

**Comment by RangerMauve** [comment]
This would be wonderful. I've been using Electron for their protocol handler APIs in [Agregore](https://agregore.mauve.moe/) but would love to switch to something a bit more rusty. 🥰

**Comment by jdm** [comment]
This is the kind of work that I'm a bit hesitant about for a couple reasons:
* I know nothing about ipfs or its use-cases. What kind of content is accessible this way that a rendering engine may need to access?
* I don't have a good sense for the maintenance burden of this over time

I'm not necessarily arguing against it, but I think I'd like to understand the value of it better.

**Reply by RangerMauve** [reply]
I think what would be best isn't even direct IPFS integration into Servo, but a nice API for adding new protocol handlers.

For example, Electron's [Protocol](https://www.electronjs.org/docs/latest/api/protocol) API makes it easier for applications (in this case Servo embedders) to adapt protocol handlers to whatever they need. I personally use it for p2p protocols, but I think a lot of applications use it just as internal links to pages or new authentication schemes.

From there we could leave the IPFS specific parts to some community experimentation (maybe Fabrice's work in Capyloon?) and if there's a clear case for it adding value to Servo itself, we could have that discussion then.

Does that sounds like a good approach?

**Reply by RangerMauve** [reply]
Also, having just the protocol handler APIs would reduce overall maintenence burden on Servo itself.

**Reply by jdm** [reply]
I would be in favour of adding an embedding API for handling new protocol handlers, for sure!

**Reply by fabricedesre** [reply]
Indeed extending the embedding interface is much better. A couple of relevant links:
- https://github.com/servo/servo/blob/8e300ca5bd6ca159992750ca3f5b281123c045f1/components/net/fetch/methods.rs#L630 where the embedder should be hooked up.
- the current embedder api to extend: https://github.com/servo/servo/blob/0d0cfd030347ab0711b3c0607a9ee07ffe7124cf/components/embedder_traits/lib.rs

I guess step 1 is to figure out the API shape.

**Reply by jdm** [reply]
To fit neatly into the existing network stack, we'll want the embedder to provide a scheme and callback, where the callback returns a `Future<Response>`. We'll want to provide arguments like the request and user agent, too. This would probably be part of [EmbedderMethods](https://github.com/servo/servo/blob/8e300ca5bd6ca159992750ca3f5b281123c045f1/components/compositing/windowing.rs#L174) today (some kind of `register_protocol` method) but it's pretty weird to have that living with the windowing integration.

**Reply by fabricedesre** [reply]
Yep, it would be better to have a new embedder trait (and move `get_user_agent_string()` to it). I tried to sketch something in https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9959176b2e2ba28afdd734753eb60213 

Not super happy with how it looks, but I can't find a way to get rid of the Box in `Box<dyn Future<Output = Response>>`.

**Reply by jdm** [reply]
Yeah, until async traits are standardized we probably can't do better than that. That looks like a good place to start!

**Reply by RangerMauve** [reply]
If I may, from having tried to add custom protocols to a few different environments (chromium, webkit, android, iOS, electron) I've found some functionality that's been important:

- Streaming interface for request and response bodies
- Being able to register schemes as "standard" in the URL handling in the stack
- Being able to register schemes as "secure" so that they can serve the same purpose as HTTPS

This is stuff that has bit me in the rear before when it wasn't available either by making it harder to tightly integrate custom URLs in web apps or made things grossly inefficient or unable to handle large amounts of data.

Adding these in at the network layer so that higher layers (like web extensions) could intercept requests is also very useful.

**Reply by fabricedesre** [reply]
Yes, Gecko also has a bunch of flags you can set on protocol handlers to decide in which context you can load a resource, and if it should be considered a secure context. I'd also like to be able to set a default CSP for a given protocol.

What do you mean by "standard" urls?

**Reply by RangerMauve** [reply]
Not sure if Servo and Gecko have this distinction, but in Chromium, marking a URL as "standard" means that it has a "hostname" portion which gets parsed out to set the origin.

https://source.chromium.org/chromium/chromium/src/+/main:url/url_util.cc;l=594?q=AddStandardScheme&ss=chromium

This is where you can configure it to parse out the user/host/port bits from the origin.

Without it origin isolation doesn't work well and URL parsing adds the hostname portion to the path which can mess with application logic using relative URLs
